Driver service layer

KsMacko opened the pull request:
---
## 🔍 Изменения:
> – Добавлен сервисный слой
> – Добавила dto для работы с оценками
> –  Возникла проблема n+1 и попробовала выборку полей в спецификации, для этого появился интерфейс (Projection) с методами доступа к этим полям
 > – Добавила dto обертку для единичных результатов (WrappedResponse)
## ❗Примечания
> Укажите, если:
> – Касательно n+1 в моей ситуации, если не писать ну очень длинный нативный запрос с объединением таблиц и подзапросом выйдет так, что придется выбрать все водителей, а потом к ним высчитывать оценки, я пыталась найти решение, чтоб вместе с фильтром (спецификацией) можно было получить все необходимое одним проходом

KsMacko committed:
service layer update

KsMacko committed:
service layer update

KsMacko committed:
service layer update
updated rate operations

IvanKrtsk commented on driver_service/src/main/java/com/internship/driver_service/entity/Rate.java:
полезно хранить дату и время с указанием временной зоны, посмотри в сторону OffsetDateTime

IvanKrtsk commented on driver_service/src/main/java/com/internship/driver_service/enums/OperationResult.java:
выглядит как будто избыточно. с точки зрения расширения может быть полезно, но не уверен, что под одно значение стоит заводить целый енам(мб оправдано использованием, но я не увидел пока нигде)

IvanKrtsk commented on driver_service/src/main/java/com/internship/driver_service/service/CommandDriverProfileService.java:
```diff
@@ -0,0 +1,72 @@
+    @Transactional
+    public ProfileDto createProfile(ProfileDto profileDto) {
+        DriverAccount driverAccount = driverAccountRepo.findById(profileDto.profileId())
+                .orElseThrow(()->new RuntimeException("driver.account.notExists"));
```
а у тебя все ошибки как RuntimeException?

antonnotdead replied:
Надо написать свои классы исключений

IvanKrtsk commented on driver_service/src/main/java/com/internship/driver_service/service/ReadDriverProfileService.java:
```diff
@@ -0,0 +1,99 @@
+                filterRequest.sortBy().getFieldName());
+        return PageRequest.of(filterRequest.page(), filterRequest.size(), sort);
+    }
+    private void validateTheId(Long id){
```
мб правильнее будет вынести в отдельный класс эти вспомогательные процедуры?
единая ответственность, если будешь расширять, легче просто заинжектить бин, в котором будет такая процедурка

KsMacko committed:
formatted code
variables moved to /env
added abstract mapper
created profile validation manager

antonnotdead commented on driver_service/.env:
.env не грузиться в репозиторий, добавь в .gitignore

antonnotdead commented on driver_service/src/main/resources/db/changelog/changeset/create-driver-tables.yaml:
информацию о машине в отдельную таблицу лучше вынести для соблюдения 3НФ

antonnotdead commented on driver_service/src/main/java/com/internship/driver_service/service/specification/DriverSpecificationService.java:
```diff
@@ -0,0 +1,65 @@
+                            cb.equal(root.get(driverStatus), statusToFilter)));
+            Optional.ofNullable(filter.fareType())
+                    .ifPresent(fareTypeToFilter -> predicates.add(
+                            cb.equal(root.get(fareType), fareTypeToFilter)));
```
Вынеси логику в отдельный метод

antonnotdead commented on driver_service/src/main/java/com/internship/driver_service/enums/OperationResult.java:
```diff
@@ -0,0 +1,11 @@
+@Getter
+public enum OperationResult {
+    SUCCESS("success");
+    private final String value;
```
Не имеет смысла создавать енам для одного значения

antonnotdead commented on driver_service/src/main/java/com/internship/driver_service/dto/WrappedResponse.java:
```diff
@@ -0,0 +1,14 @@
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class WrappedResponse<T> {
```
Сервисы должны возвращать либо entity либо dto

KsMacko committed:
added env to gitignore
added car entity, dto and validation manager
updated specification

KsMacko committed:
some changes

DaniilTereshko commented on driver_service/src/main/java/com/internship/driver_service/dto/mapper/CarMapper.java:
```diff
@@ -0,0 +1,19 @@
+public interface CarMapper extends AbstractMapper<CarDto, Car>{
+    CarMapper converter = Mappers.getMapper(CarMapper.class);
+
+    Car handleDto(CarDto carDto);
+    CarDto handleEntity(Car car);
```
у тебя эти методы есть в AbstractMapper

DaniilTereshko commented on driver_service/src/main/java/com/internship/driver_service/dto/transfer_objects/DriverFilterRequest.java:
```diff
@@ -1,4 +1,4 @@
-package com.internship.driver_service.dto;
+package com.internship.driver_service.dto.transfer_objects;
```
нижние подчеркивания не используем в нейминге пакетов, качается и driver_service и всех остальных

DaniilTereshko commented on driver_service/src/main/java/com/internship/driver_service/utils/ProfileValidationManager.java:
```diff
@@ -0,0 +1,25 @@
+    private final DriverProfileRepo driverProfileRepo;
+    public void checkIfProfileExists(Long id) {
+        if(driverProfileRepo.findById(id).isEmpty())
+            throw new RuntimeException("driver.notFound");
```
кастомные эксепшены, ключи в константы

DaniilTereshko commented on driver_service/src/main/java/com/internship/driver_service/utils/ProfileValidationManager.java:
```diff
@@ -0,0 +1,25 @@
+    }
+    public DriverProfile getDriverProfile(Long id) {
+        return driverProfileRepo.findById(id)
+                .orElseThrow(()->new RuntimeException("driver.notFound"));
```
отформатируй

DaniilTereshko commented on driver_service/src/main/java/com/internship/driver_service/utils/ProfileValidationManager.java:
отформатируй, делай перед комитом во всех файлах ctrL+alt+O , Ctrl + Alt + L,
либо для форматрования правой кнопкой на самый корневой пакет - refotmat code + optimize import

KsMacko committed:
reformatted code
